### PR TITLE
SQLExtension: Fix show reference command with use reference@hash

### PR DESCRIPTION
**Pull request by ajantha-bhat**
Whenever SQL "USE REFERENCE @ HASH" is used, It only updates the [NessieIcebergClient](https://github.com/apache/iceberg/blob/master/nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java) 's `updatableReference`. So, there is no change in the actual Nessie reference in the backend.

But when current reference info is fetched, It fetches from the backend which is not aware of "USE REFERENCE @ HASH".  So, this is confusing for the users and also it is not the current state of the reference.  Hence, this fix.

Also, assign ref, create ref from, and merge ref will honor the value set by "use reference @hash" now.

### CODE REVIEW

File: clients/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java [hunk 1]
```diff
@@ -374,7 +374,7 @@ void useShowReferencesAtTimestampWithTimeZone(DateTimeFormatter dateTimeFormatte
     assertThat(sql("USE REFERENCE %s AT `%s` IN nessie ", refName, timeWithZone))
         .containsExactly(row("Branch", refName, lastCommitBeforeTimePredicate.getHash()));
     assertThat(sql("SHOW REFERENCE IN nessie"))
-        .containsExactly(row("Branch", refName, lastRefHashGlobally));
+        .containsExactly(row("Branch", refName, lastCommitBeforeTimePredicate.getHash()));
```
Comment (snazy) [inline clients/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java:377]: Not sure why this change is necessary, because the logic for timestamps shouldn't change, right?
And there's no more check against the reason mentioned in the comment - and the variable is now unused.

File: clients/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala [hunk 2]
```diff
@@ -246,7 +247,16 @@ object NessieUtils {
   ): Reference = {
     val refName = getCurrentRefName(currentCatalog, catalog)
```
Comment (snazy) [inline clients/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala:248]: Why not update the existing getCurrentRefName to return a tuple of name + optional-hash?

File: clients/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java (hunk 1 above)
Comment (ajantha-bhat) [reply clients/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java:377]: > Not sure why this change is necessary, because the logic for timestamps shouldn't change, right?

`USE ref AT hash or timestamp` has the same behaviour (just that timestamp will be mapped to a hash -> uses `NessieUtils#calculateRef`). 

Without this PR, after setting `USE reference at oldHashTimestamp`, show reference was always giving the head commit hash of that reference. Hence, need to modify it now as the show reference will return the current hash under use.

> and the variable is now unused.

Oh, I will remove it.

File: clients/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala (hunk 2 above)
Comment (ajantha-bhat) [reply clients/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala]: originally thought of changing but didn't change as I had to explore some Scala syntaxes if I change.

Updated now.

**Review by snazy — APPROVED** [review]
